lib: cmsis_rtos_v2: Additional CMSIS RTOS v2 Support

carlos-stuart opened the pull request:
Extended the current CMSIS RTOS v2 support by adding the ability for memory pools to be created dynamically on the heap using default parameters (in the same way message queues were previously implemented) and added a way for thread stacks to be _dynamically_ created (and allow default parameters when no attributes are passed in).

Thread stacks are actually not dynamically created (they are defined in a _CMSIS_V2_THREAD_DYNAMIC_MAX_COUNT_ sized array), they are however dynamically assigned to threads when a stack buffer is not passed in via the optional attributes.

Additionally I've fixed an issue where one more CMSIS thread could be created than the maximum defined by the Kconfig variable - this happened because of an assert doing a less than or equal comparison where it needed to do a less than due to when the counter was incremented.

I've also changed the way the thread counter is incremented to be slightly safer (chances are the compiler would have optimised it to be safe before) by assigning the return value from atomic_inc rather than assigning the value before and then using atomic_inc.

I've also uncrustified all the related source and header files.

I've also added an implied dependency (from the source) to the Kconfig file.

No one is listed as the code owner (in the CODEOWNERS file) for this library so I am unsure who I should tag for a review.

carlos-stuart commented:
@nashif Please would you consider reviewing this pull request? Many thanks!

nashif commented:
@carlos-stuart Thanks for the PR, will take a look.

carlos-stuart commented:
Pushed a minor change to the Kconfig commit. The heap is not required to be non-zero now thread, memory pool and message queue have Kconfig variables to configure their respective maximum sizes, which are defaulted to 0 i.e. not requiring any heap. Additional the scheduler was listed as a dependency when in fact there could be multiple valid schedulers.

carlos-stuart commented:
Another minor change; changed the thread pool array definition to use the _K_THREAD_STACK_ARRAY_DEFINE_ macro rather than the normal _K_THREAD_STACK_DEFINE_ macro,

rgundi reviewed (changes requested):
Thanks for the PR. Could you please split commits e2fdad7f3119b939efe627f44085a1fe4cf975f9 and 948bc56edebaffdf6dc7b569e6c01f46f850417f into two? It's better to separate the implementation and tests into different commits from a review perspective.

carlos-stuart commented:
> Thanks for the PR. Could you please split commits [e2fdad7](https://github.com/zephyrproject-rtos/zephyr/commit/e2fdad7f3119b939efe627f44085a1fe4cf975f9) and [948bc56](https://github.com/zephyrproject-rtos/zephyr/commit/948bc56edebaffdf6dc7b569e6c01f46f850417f) into two? It's better to separate the implementation and tests into different commits from a review perspective.

@rgundi I've split those commits as requested; I also used the opportunity to rebase on to master.

carlos-stuart commented:
Addressed a newly identified issue whereby a segmentation fault (attempt to read from an invalid address) would occur when an optional attribute structure (attr) was passed into a CMSIS RTOS v2 new function with an invalid/NULL name. This was caused by a memcpy that copied the name into another buffer, but if the name pointed to an invalid address this would then fail.

This has been fixed by checking if the name is NULL and if it is then using the default name from the init structure.

This fix has been applied to all the relevant CMSIS RTOS v2 functions.

rgundi commented on tests/cmsis_rtos_v2/src/thread_apis.c:
```diff
@@ -29,6 +30,80 @@ static osThreadAttr_t thread2_attr = {
 	.priority = osPriorityHigh,
 };
 
+static void thread1_dynamic(void *argument)
```
The body of the newly introduced functions are almost same as the existing functions. It is better to use the existing functions itself and distinguish between dynamic and regular threads using arguments taken by the thread functions.

carlos-stuart replied:
Sure - I'll pass pointers to the globals in as thread arguments.

rgundi reviewed (changes requested):
Thanks for the PR. It looks mostly good. There are a few review comments which i feel you should address though.

carlos-stuart commented on lib/cmsis_rtos_v2/thread.c:
```diff
@@ -83,36 +100,69 @@ osThreadId_t osThreadNew(osThreadFunc_t threadfunc, void *arg,
+		return NULL;
+	}
+
+	if (thread_num >= CONFIG_CMSIS_V2_THREAD_MAX_COUNT) {
```
The purpose of this test is that the __ASSERT only fails on a system that have those enabled, without this check and returning NULL more threads than the maximum can be created. I'd say it's safer to have the always-enabled runtime check than relying on the __ASSERT alone. What do you think?

rgundi replied:
Yes, you are correct. I use ASSERTs so much that i have taken them for granted. :(
Since this is a condition which essentially means there is no point in proceeding further, i believe we should use only the if statement and remove __ASSERT. Is it possible for you to scrub the source code of CMSIS RTOS v2 for such usages and correct it?

carlos-stuart commented on lib/cmsis_rtos_v2/thread.c:
```diff
@@ -83,36 +100,69 @@ osThreadId_t osThreadNew(osThreadFunc_t threadfunc, void *arg,
+		__ASSERT(thread_num < CONFIG_CMSIS_V2_THREAD_MAX_COUNT,
+			 "exceeded maximum dynamic thread count");
+		return NULL;
+	}
```
This should be checking the dynamic thread count. Will update.

carlos-stuart commented on lib/cmsis_rtos_v2/wrapper.h:
```diff
@@ -19,6 +19,7 @@ struct cv2_thread {
 	struct k_poll_signal poll_signal;
 	struct k_poll_event poll_event;
 	u32_t signal_results;
+	void *stack;
```
It's not, you're right. It just feels wrong not to keep track of it in the struct, but it serves no purpose, I'll drop it.

carlos-stuart commented on tests/cmsis_rtos_v2/src/mempool.c:
```diff
@@ -26,6 +26,88 @@ static const osMemoryPoolAttr_t mp_attrs = {
+	mp_id = osMemoryPoolNew(MAX_BLOCKS, sizeof(struct mem_block),
+				&mp_attrs);
+	zassert_true(mp_id != NULL, "mempool creation failed");
+
```
I think they can have a common body. I'll update that.

carlos-stuart commented on lib/cmsis_rtos_v2/thread.c:
```diff
@@ -83,36 +100,69 @@ osThreadId_t osThreadNew(osThreadFunc_t threadfunc, void *arg,
-	__ASSERT(attr != NULL,
-		 "Zephyr requirement: Pass attributes explicitly for ThreadNew\n");
+	__ASSERT((attr->stack_mem == NULL) || (attr->stack_size > 0),
+		 "stack_size must be greater than 0 if stack_mem is not NULL\n");
```
The original `__ASSERT` is as intended.

It is stating that if `attr->stack_mem` is `not NULL` then `attr->stack_size` must be greater than 0.

The truth table would be:

```
stack_mem    stack_size    result?

  NULL           0          true
  NULL           >0         true
 Not NULL        0          false
 Not NULL        >0         true
````

From the above truth table the following boolean expression can be derived:

`state_mem == NULL` **OR** `stack_size > 0`

The only other way to do this would be to break it up into nested if-statements.

Opinion?

rgundi replied:
Checking for attr->stack_size when attr->stack_mem is NULL doesn't make any sense as we don't care about it.  So, I don't think you should consider the first 2 rows in your truth table. This then boils down to the condition I specified. Moreover, attr->stack_mem is being checked for NULL in "if else" at line 162 and 168 and is hence covered.

rgundi replied:
> However the condition you specified wouldn't hold true when `stack_mem` is `NULL` thus causing the `ASSERT` to be `false` and halting execution even though the attributes are valid. Logical **OR** is a lazy operator so only the first operand is evaluated if it is `true`, meaning the `stack_size` is then never checked if `stack_mem` is `NULL`.
> 
> While `stack_mem` is checked for being `NULL` later on in the function this is only to ascertain whether dynamic allocation should be attempted or not, it doesn't protect again nor detect the situation where a valid `stack_mem` pointer has been passed in but the `stack_size` is invalid i.e. 0.
> 
> Maybe a clearer implementation would be:
> 
> ```
> if (attr->stack_mem != NULL) {
>     if (attr->stack_size == 0) {
>         return NULL; /* stack_size must be non-zero if stack_mem is set. */
>     }
> }
I agree. Please go ahead with this implementation
> ```
> This more explicitly avoids any check of `stack_size` if `stack_mem` is `NULL` and still causes the function to fail (and without an `ASSERT` as suggested in another comment).

carlos-stuart commented:
I was thinking it would be fairly trivial to implement `osThreadJoin` - are you aware of a reason it wasn't implemented before? I realise the Zephyr kernel API doesn't include a join itself but as there is a wrapper function for the thread function it's fairly easy to set a flag and release a mutex so any threads waiting for this thread to join can then continue.

rgundi commented on tests/cmsis_rtos_v2/src/thread_apis.c:
```diff
@@ -169,41 +208,50 @@ static void thread3(void *argument)
 	zassert_true(id3 != NULL, "Failed creating thread3");
 
-	/* Keep delaying 10 milliseconds to ensure thread3 is done with
+	/* Keep delaying 10 milliseconds to ensure thread3_dynamic is done with
```
This should remain thread3.

rgundi reviewed (changes requested):
Please fix the checkpatch issues when you resubmit.

rgundi commented:
> I was thinking it would be fairly trivial to implement `osThreadJoin` - are you aware of a reason it wasn't implemented before? I realise the Zephyr kernel API doesn't include a join itself but as there is a wrapper function for the thread function it's fairly easy to set a flag and release a mutex so any threads waiting for this thread to join can then continue.

Well.. no real reason for not implementing it. This was the first cut and we wanted to get some feedback before proceeding further. Please go ahead and do it in a fresh PR if you are interested. There are also some more things not implemented which you can look at (doc/porting/cmsis_rtos_v2.rst).

carlos-stuart committed:
lib: cmsis_rtos_v2: Uncrustify source files

Ran uncrustify on all library source files to ensure a compliant base to
work from.

Signed-off-by: Carlos Stuart <carlosstuart1970@gmail.com>

carlos-stuart committed:
tests: cmsis_rtos_v2: Uncrustify source files

Ran uncrustify on all test source files to ensure a compliant base to
work from.

Signed-off-by: Carlos Stuart <carlosstuart1970@gmail.com>

carlos-stuart committed:
lib: cmsis_rtos_v2: Additional Kconfig dependency

Added Kconfig dependency that NUM_PREEMPT_PRIORITIES must be at least
osPriorityISR (56). This was enforced by a build assert message but not
decribed in the Kconfig.

Signed-off-by: Carlos Stuart <carlosstuart1970@gmail.com>

carlos-stuart committed:
lib: cmsis_rtos_v2: Message queue checks

Added some additional checks when creating a message queue to ensure the
size of the queue does not exceed the size of the buffer passed in via
the optional attributes.

Added a new Kconfig option to limit the maximum size of a message queue
dynamically allocated on the heap.

Added a check to ensure the heap is at least large enough to hold a
maximum size dynamically allocated queue.

Signed-off-by: Carlos Stuart <carlosstuart1970@gmail.com>

carlos-stuart committed:
lib: cmsis_rtos_v2: Dynamic memory pools

Implemented dynamic allocation of memory pools in a similar to manner to
what was already implemented for message queues. Added all the same
checks on size vs. maximum allowed and current heap.

Added an additional Kconfig variable to define the maximum size of a
dynamically allocated memory pool.

Signed-off-by: Carlos Stuart <carlosstuart1970@gmail.com>

carlos-stuart committed:
tests: cmsis_rtos_v2: Dynamic memory pool tests

Implemented a second set of memory pool tests that function on a
dynamically allocated memory pool. Abstracted some of the common
functionality to a common function.

Signed-off-by: Carlos Stuart <carlosstuart1970@gmail.com>

carlos-stuart committed:
lib: cmsis_rtos_v2: Dynamic thread stacks

Implemented dynamic thread stacks for CMSIS threads by declaring an
array of default sized thread stacks. Allocation cannot be done on the
heap as some architectures require strict alignment for stacks so the
macro must be used to define the stack to ensure most compatibility.

Added a Kconfig variable to limit the number of dynamic threads on the
system (they also count towards total CMSIS thread count). This is so a
developer can have fine grained control over how many dynamic threads
can be allocated because all their stacks must be allocated up front so
could use a lot of memory needlessly if oversubscribed. The default
value is 0 which effectively disabled dynamic threads but also reduces
the memory impact to almost none.

Fixed an assert bug where thread_num was being tested against the
maximum allowed CMSIS threads - it previous checked for less than or
equal which actually (due to when the increment happens) allowed there
to be one more thread. The check now correctly uses less than and only
allowed up to the defined maximum.

Signed-off-by: Carlos Stuart <carlosstuart1970@gmail.com>

carlos-stuart committed:
tests: cmsis_rtos_v2: Dynamic thread stack tests

Added a set of tests that use the default thread stacks to ensure they
work the same as the passed in stacks. Abstracted the common
functionality into functions.

Signed-off-by: Carlos Stuart <carlosstuart1970@gmail.com>

carlos-stuart committed:
lib: cmsis_rtos_v2: Default name if name is NULL

Fixed an issue whereby if an attribute structure was passed into a CMSIS
RTOS v2 'new' function with an invalid address i.e. NULL assigned to the
name (char*) member the memcpy at the end of each new function
would cause a segmentation fault i.e. read from an invalid
address.

This has been fixed by checking if the name is NULL and using the
default name from the init struct if it is. This is the same name
that would be used if not passing in the optional attr function
argument.

Changed the memcpy to strncpy to ensure that the copy does not read
beyond the end of the source string and changed the length from 16 to 15
(by means of a `sizeof(...)-1`) of the destination buffer to ensure that
it will always be nul-terminated.

Signed-off-by: Carlos Stuart <carlosstuart1970@gmail.com>

carlos-stuart committed:
lib: cmsis_rtos_v2: Default thread prioity

If an unitialized/zeroed optional attribute was passed to osThreadNew
the priority would be osThreadNone i.e. uninitialized. This causes an
ASSERT to be hit as the priority isn't valid (it is not between
osPriorityIdle and osPriorityISR).

The fix checks the passed in priority is not osPriorityNone and assigns
osPriorityNormal. This is the correct CMSIS behaviour.

The ASSERT will still be hit if the priority is invalid (<0).

Signed-off-by: Carlos Stuart <carlosstuart1970@gmail.com>

carlos-stuart commented:
> > I was thinking it would be fairly trivial to implement `osThreadJoin` - are you aware of a reason it wasn't implemented before? I realise the Zephyr kernel API doesn't include a join itself but as there is a wrapper function for the thread function it's fairly easy to set a flag and release a mutex so any threads waiting for this thread to join can then continue.
> 
> Well.. no real reason for not implementing it. This was the first cut and we wanted to get some feedback before proceeding further. Please go ahead and do it in a fresh PR if you are interested. There are also some more things not implemented which you can look at (doc/porting/cmsis_rtos_v2.rst).

Thanks, I'll implement `osThreadJoin` and `osThreadDetach` in another PR. I'll have a look at the docs and ensure they're updated in that PR too to reflect those changes.

Will also see if there are any others features I can implement. Thanks for reviewing this PR, your help is really appreciated.

carlos-stuart commented:
Just rebased on to `master`.

carlos-stuart commented:
> Please fix the checkpatch issues when you resubmit.

Done. I still get some locally when I run it as a pre-commit hook however the CI doesn't complain about those. They were warnings and didn't really make a lot of sense (resolving them either caused another different warning or made the code violate the coding style (which `uncrustify` then just reverted).

rgundi reviewed (approved):
LGTM.
